File: python/schemapin/discovery.py
"""Public key discovery via .well-known URIs per RFC 8615."""

import json
from typing import Dict, Any, Optional
from urllib.parse import urljoin
import requests


class PublicKeyDiscovery:
    """Handles public key discovery from .well-known endpoints."""

    @staticmethod
    def construct_well_known_url(domain: str) -> str:
        """
        Construct .well-known URI for SchemaPin public key discovery.
        
        Args:
            domain: Tool provider domain
            
        Returns:
            Full .well-known URI
        """
        if not domain.startswith(('http://', 'https://')):
            domain = f"https://{domain}"
        return urljoin(domain, '/.well-known/schemapin.json')

    @staticmethod
    def validate_well_known_response(response_data: Dict[str, Any]) -> bool:
        """
        Validate .well-known response structure.
        
        Args:
            response_data: Parsed JSON response
            
        Returns:
            True if response is valid, False otherwise
        """
        required_fields = ['schema_version', 'public_key_pem']
        return all(field in response_data for field in required_fields)

    @classmethod
    def fetch_well_known(cls, domain: str, timeout: int = 10) -> Optional[Dict[str, Any]]:
        """
        Fetch and validate .well-known/schemapin.json from domain.
        
        Args:
            domain: Tool provider domain
            timeout: Request timeout in seconds
            
        Returns:
            Parsed response data if valid, None otherwise
        """
        try:
            url = cls.construct_well_known_url(domain)
            response = requests.get(url, timeout=timeout)
            response.raise_for_status()
            
            data = response.json()
            if cls.validate_well_known_response(data):
                return data
            return None
            
        except (requests.RequestException, json.JSONDecodeError, ValueError):
            return None

    @classmethod
    def get_public_key_pem(cls, domain: str, timeout: int = 10) -> Optional[str]:
        """
        Get public key PEM from domain's .well-known endpoint.
        
        Args:
            domain: Tool provider domain
            timeout: Request timeout in seconds
            
        Returns:
            PEM-encoded public key if found, None otherwise
        """
        well_known_data = cls.fetch_well_known(domain, timeout)
        if well_known_data:
            return well_known_data.get('public_key_pem')
        return None

    @classmethod
    def get_developer_info(cls, domain: str, timeout: int = 10) -> Optional[Dict[str, str]]:
        """
        Get developer information from .well-known endpoint.
        
        Args:
            domain: Tool provider domain
            timeout: Request timeout in seconds
            
        Returns:
            Dictionary with developer info if available, None otherwise
        """
        well_known_data = cls.fetch_well_known(domain, timeout)
        if well_known_data:
            return {
                'developer_name': well_known_data.get('developer_name', 'Unknown'),
                'schema_version': well_known_data.get('schema_version', '1.0'),
                'contact': well_known_data.get('contact', ''),
            }
        return None

File: go/pkg/discovery/discovery.go
// Package discovery provides .well-known endpoint discovery for SchemaPin.
package discovery

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// WellKnownResponse represents .well-known/schemapin.json structure
type WellKnownResponse struct {
	SchemaVersion string   `json:"schema_version"`
	DeveloperName string   `json:"developer_name"`
	PublicKeyPEM  string   `json:"public_key_pem"`
	Contact       string   `json:"contact,omitempty"`
	RevokedKeys   []string `json:"revoked_keys,omitempty"`
}

// PublicKeyDiscovery handles .well-known endpoint discovery
type PublicKeyDiscovery struct {
	client *http.Client
}

// NewPublicKeyDiscovery creates a new PublicKeyDiscovery instance
func NewPublicKeyDiscovery() *PublicKeyDiscovery {
	return &PublicKeyDiscovery{
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// ConstructWellKnownURL constructs the .well-known URL for a domain
func (p *PublicKeyDiscovery) ConstructWellKnownURL(domain string) string {
	return fmt.Sprintf("https://%s/.well-known/schemapin.json", domain)
}

// FetchWellKnown fetches and parses the .well-known/schemapin.json file
func (p *PublicKeyDiscovery) FetchWellKnown(ctx context.Context, domain string) (*WellKnownResponse, error) {
	url := p.ConstructWellKnownURL(domain)

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch .well-known file: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	var wellKnown WellKnownResponse
	if err := json.NewDecoder(resp.Body).Decode(&wellKnown); err != nil {
		return nil, fmt.Errorf("failed to decode .well-known response: %w", err)
	}

	return &wellKnown, nil
}

// GetPublicKeyPEM retrieves the public key PEM from .well-known endpoint
func (p *PublicKeyDiscovery) GetPublicKeyPEM(ctx context.Context, domain string) (string, error) {
	wellKnown, err := p.FetchWellKnown(ctx, domain)
	if err != nil {
		return "", err
	}

	if wellKnown.PublicKeyPEM == "" {
		return "", fmt.Errorf("no public key found in .well-known response")
	}

	return wellKnown.PublicKeyPEM, nil
}

// ValidateKeyNotRevoked checks if a public key is not in the revoked keys list
func (p *PublicKeyDiscovery) ValidateKeyNotRevoked(ctx context.Context, publicKeyPEM, domain string) (bool, error) {
	wellKnown, err := p.FetchWellKnown(ctx, domain)
	if err != nil {
		return false, err
	}

	// Check if the key is in the revoked list
	for _, revokedKey := range wellKnown.RevokedKeys {
		if revokedKey == publicKeyPEM {
			return false, nil
		}
	}

	return true, nil
}

// GetDeveloperInfo retrieves developer information from .well-known endpoint
func (p *PublicKeyDiscovery) GetDeveloperInfo(ctx context.Context, domain string) (map[string]string, error) {
	wellKnown, err := p.FetchWellKnown(ctx, domain)
	if err != nil {
		return nil, err
	}

	info := map[string]string{
		"developer_name": wellKnown.DeveloperName,
		"schema_version": wellKnown.SchemaVersion,
		"domain":         domain,
	}

	if wellKnown.Contact != "" {
		info["contact"] = wellKnown.Contact
	}

	return info, nil
}
